chore: Updated content-type header parsing

jsumners opened the pull request:
Resolves #6332.

This is a follow-on from #6303. This introduces a more robust content-type header value parsing algorithm.

***IMPORTANT:*** this results in much stricter content-type handling. Prior to this PR, we were accepting, and handling, headers like `content-type: a-simple-string`. That is not valid. There must be a type **and** a subtype present. So a "valid" form would be `content-type: a-simple-string/foo`.

This change _also_ enforces normalization of the content-type value. For example, if the request has `application/json; charset=utf-8` as the value, it will be normalized to `application/json; charset="utf-8"`.

jsumners committed:
chore: Updated content-type header parsing

gurgunday reviewed (commented):
I think this is hard to land in v5

Also, we should check how much it affects the performance

Unfortunately we removed the benchmark CI, someone should run it locally (I'll do it when I review)

jsumners commented:
> It like a light weight version of content-type parser? Removed the parameters parts, it is ok on our case because we never use it.

In our code base, "content-type parser" refers to the thing that collects and manages the functions used to parse HTTP POST bodies. Prior to this PR, the indicator used to denote what format those POST bodies is in was inspected in various ways that are very fragile. This PR introduces a dedicated identifier parser.

> How the performance after the change?

I don't know. Likely slower than the previous fragile checks, but likely not the slowest possible. Unfortunately, proper handling of the `content-type` header value is an imperative. We cannot continue using the previous fragile inspections.

mcollina commented on lib/content-type.js:
```diff
@@ -0,0 +1,110 @@
+      return
+    }
+
+    parts = parts[1].split(';')
```
Here we are splitting again. Can we just do it in one pass?

mcollina reviewed (changes requested):
Are you sure that there are no content types without a `/`? I think a user could create one like that, and I feel we should kind of support it.

jsumners commented:
> Are you sure that there are no content types without a `/`? I think a user could create one like that, and I feel we should kind of support it.

Please review https://httpwg.org/specs/rfc9110.html#media.type

The required format is `type/subtype`. Anything written as `type` is not spec compliant. We opt to adhere to the specs: https://github.com/fastify/fastify/blob/main/docs/Reference/Principles.md#specification-adherence

jsumners commented on lib/content-type.js:
```diff
@@ -0,0 +1,110 @@
+
+    const hv = headerValue.trim()
+    let parts = (() => {
+      const [first, ...rest] = hv.split('/')
```
I'll construct a benchmark on this. I'm not convinced this is slower than a regex for the use case.

jsumners commented on lib/content-type.js:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    for (let i = 1; i < parts.length; i += 1) {
+      const [key, value] = parts[i].trim().split('=')
```
How else can we handle a list of key+value pairs?

jsumners replied:
Writing a regular expression to match the parameters has definitely entered nasty territory, exactly what I was trying to avoid. I don't mind fairly simple/strait forward regular expressions, but I detest ones that need a novel to explain them.

jsumners commented:
@mcollina I have written the benchmark below to test the performance of splitting out only the `type` and `subtype` parts. I agree that the IIFE is too expensive. It looks like the regular expression version becomes worth it as we start dealing with the parameters.

<details>
<summary>Benchmark</summary>

```js
'use strict'

const assert = require('node:assert')
const { Suite } = require('bench-node')
const suite = new Suite({
  reporter
})

const basicMediaType = 'application/json'
const paramsMediaType = 'application/json; charset=utf-8; foo=bar'

suite.add('iife split (basic)', () => {
  let parts = (() => {
    const [first, ...rest] = basicMediaType.split('/')
    if (rest.length === 0) return [first]
    return [first, rest.join('/')]
  })()
  assert.equal(parts[0], 'application')
  assert.equal(parts[1], 'json')
})

suite.add('iife split (params)', () => {
  let parts = (() => {
    const [first, ...rest] = paramsMediaType.split('/')
    if (rest.length === 0) return [first]
    return [first, rest.join('/')]
  })()

  const type = parts[0]
  parts = parts[1].split(';')
  const subtype = parts[0]
  
  assert.equal(type, 'application')
  assert.equal(subtype, 'json')
})

suite.add('long split (basic)', () => {
  let parts = basicMediaType.split('/')
  assert.equal(parts[0], 'application')
  assert.equal(parts[1], 'json')
})

suite.add('long split (params)', () => {
  let parts = paramsMediaType.split('/')
  const type = parts[0]

  parts = parts[1].split(';')
  const subtype = parts[0]

  assert.equal(type, 'application')
  assert.equal(subtype, 'json')
})

const tokensReg = /^([\w!#$%&'*+.^`|~-]+)\/([\w!#$%&'*+.^`|~-]+)\s*;?/
suite.add('regex split (basic)', () => {
  const matches = tokensReg.exec(basicMediaType)
  const type = matches[1]
  const subtype = matches[2]
  assert.equal(type, 'application')
  assert.equal(subtype, 'json')
})

suite.add('regex split (params)', () => {
  const matches = tokensReg.exec(paramsMediaType)
  const type = matches[1]
  const subtype = matches[2]
  assert.equal(type, 'application')
  assert.equal(subtype, 'json')
})

suite.run()

const { format } = new Intl.NumberFormat('en-US', {
  useGrouping: true
})
function reporter (results) {

  for (const result of results) {
    const ops = format(result.opsSec)
    console.log(`${result.name}: ${ops} ops/sec`)
  }
}
```

</details>

```
❯ node --allow-natives-syntax index.js
iife split (basic): 9,655,587.18 ops/sec
iife split (params): 6,261,203.487 ops/sec
long split (basic): 15,462,605.433 ops/sec
long split (params): 8,142,902.648 ops/sec
regex split (basic): 15,020,238.273 ops/sec
regex split (params): 14,836,393.537 ops/sec
```

Eomm commented on lib/content-type.js:
```diff
@@ -0,0 +1,110 @@
+      return [first, rest.join('/')]
+    })()
+
+    if (parts.length !== 2) {
```
```suggestion
    if (parts.length !== 2) {
      // This is an invalid content-type
```

Eomm reviewed (commented):
Closes https://github.com/fastify/fastify/pull/6077

Minor suggestions

jsumners commented:
I'll be getting back to this soon.

jsumners committed:
address feedback

jsumners commented on test/schema-validation.test.js:
```diff
@@ -1416,124 +1416,157 @@ test('Schema validation will not be bypass by different content type', async t =
   })
-  t.assert.strictEqual(invalid4.status, 400)
-  t.assert.strictEqual((await invalid4.json()).code, 'FST_ERR_VALIDATION')
+  t.assert.strictEqual(found.status, 415)
```
I'm not sure the regular expression based algorithm is going to work. The current `tokensReg` is not matching this test correctly. I am unable to, at least at the moment, to determine a tweak to the expression that will work. The whole point of this PR is to cover these invalid header values such that the schema validation will not be skipped. The assertion for a `415` is correct.

mcollina replied:
Is there any ohter viable approach?

jsumners committed:
refactor algorithm

jsumners committed:
Update lib/content-type-parser.js

Co-authored-by: Manuel Spigolon <behemoth89@gmail.com>
Signed-off-by: James Sumners <321201+jsumners@users.noreply.github.com>

jsumners committed:
appease coverage

mcollina reviewed (approved):
lgtm

jean-michelet reviewed (approved):
Nice!

Let's close this after merge:
https://github.com/fastify/fastify/issues/6332

gurgunday reviewed (commented):
Are we landing this as a bug fix? Wouldn't SemVer major be more appropriate?

danielgindi commented:
This PR is a breaking change actually.

Prior to this one, Content-Type was not strictly validated (btw it is not strictly validated in most http server libs).

Now let's say that we are expecting some malformed header from a hard coded client we cannot control - and we have registered a custom content type parser.
- If we register it as a string - it throws immediately.
- If we register it as a regex - it registers, but throws 415 for the actual request, as `isValid` is called.

This library does not allow any workaround, and forces us to replace the library or stick to v5.7.x.
As there is no way to say "Yes, I want to allow invalid Content-Type headers" or at least - match through the registered content type parsers - without checking for `isValid` first.

Is this going to be adjusted?

jsumners commented:
> > Are you sure that there are no content types without a `/`? I think a user could create one like that, and I feel we should kind of support it.
> 
> Please review https://httpwg.org/specs/rfc9110.html#media.type
> 
> The required format is `type/subtype`. Anything written as `type` is not spec compliant. We opt to adhere to the specs: https://github.com/fastify/fastify/blob/main/docs/Reference/Principles.md#specification-adherence

bcomnes commented:
Whats the appropriate workaround/approach to take when a client is serving up invalid headers?

climba03003 commented:
> Whats the appropriate workaround/approach to take when a client is serving up invalid headers?

It is not clear to me `client` means `request` or `reply`?
For `request`, the behavior should be similar as before. 
Parsing do not happen when the `Content-Type` do not match the `media type` part.
If you want to ensure do something on the invalid media type, you can
```js
fastify.post('/', async function(request, reply) {
  if(request.mediaType == undefined) {
    // You can use hook to do the same thing.
    throw Error('Invalid Content-Type')
  } else {
    reply.send({ ok: true })
  }
})
```

For `response`, developer should always ensure they send a valid header.

bcomnes commented:
I'm pretty confused, and I think maybe this isn't quite right yet.

RFC 9110 states:

> “Unless noted otherwise, a recipient MAY attempt to recover a usable protocol element from an invalid construct.”

In a Fastify service that receives webhooks from Jira, for example, we observe `Content-Type` headers like:

```text
"application/json,application/json"                  (rejected after this change)
"application/json; charset=utf-8,application/json"   (still accepted after this change)
```

Best I can tell, a request ends up with duplicate `Content-Type` headers and then some intermediary collapses them into a single header. By the time they reach Fastify, they are in this weird state.

We do not appear to enable `joinDuplicateHeaders`, so I suspect something in front of the service (either in the thing sending it or our infra).  Either way, these are headers our service receives and cannot control. It might even be outside the client’s control if the request passes through other layers that introduce this behavior.

Previously, for `application/json,application/json`, I could add an encapsulated RegExp `addContentTypeParser` on the affected routes. Fastify 5.5 passed the raw header value through parser matching, our RegExp recognized the known malformed form, and we deliberately selected the JSON body parser. That seems like a good example of the application-specific recovery permitted by RFC 9110.

After this change, however, `application/json,application/json` is rejected before my custom parser is consulted. I instead have to handle it in an `onRequest` hook by parsing and mutating the header into a valid value before Fastify performs its validation.

I understand that `addContentTypeParser` manages body-parser selection rather than being a general-purpose header parser. However, registering a RegExp parser was previously a clear way to say, “For this known external header form, parse the body as FOO”

Is `onRequest` header mutation really the intended layer for recovering correctable external input? This seems like both a spec-conformance fix and a Fastify policy and extension-point change. I am open to being convinced otherwise, but this feels like behavior that should have been introduced behind an option or in a breaking release.

@climba03003, if my understanding is correct, checking `request.mediaType` in the handler would not work. The request handler never runs because Fastify responds with `415 Unsupported Media Type` before entering it. When inspected from an `onRequest` hook, `request.mediaType` also appears to be `"/"` for this invalid value rather than `undefined`.

My main concerns are:

- This broke a previously working use of an explicitly registered RegExp content-type parser, but it was released as a bug fix.
- Stricter default validation may be appropriate, but preventing a registered parser from implementing an application-specific recovery policy is a separate behavioral change.
- `addContentTypeParser` was a reasonably clear place to recover body-parser selection for a known malformed header. Why require applications to move that logic into `onRequest` and mutate the raw header instead?

bcomnes commented:
I sketched out my idea here: https://github.com/fastify/fastify/pull/6918 @jsumners I think this actually still aligns with the intention here, while also making this less of a breaking change. Essentially it restores the ability to use a regxp `addContentTypeParser` to correct for bad headers, and doesn't force you to move this to `onRequest` while keeping the stricter header validation.

bcomnes commented:
I also put together a WHATWG token based parsing alternative, which retains the stronger validation of this PR, but uses a more forgiving and standards based approach to invalid header recovery. Please check it out https://github.com/fastify/fastify/pull/6919

It complements https://github.com/fastify/fastify/pull/6918 which restores the location of header validation.

jsumners commented:
> RFC 9110 states:
> 
> > “Unless noted otherwise, a recipient MAY attempt to recover a usable protocol element from an invalid construct.”

"MAY" is not "MUST". As noted elsewhere, the Fastify project opts to strictly adhere to the RFCs.

> I instead have to handle it in an `onRequest` hook by parsing and mutating the header into a valid value before Fastify performs its validation.

This sounds like you are able to handle these absolutely incorrect header values.

> Is `onRequest` header mutation really the intended layer for recovering correctable external input? This seems like both a spec-conformance fix and a Fastify policy and extension-point change. I am open to being convinced otherwise, but this feels like behavior that should have been introduced behind an option or in a breaking release.

My position is: yes, the `onRequest` hook is the feature to use if you need to massage the request before it is processed.

The simple fact of the matter is that invalid content-type headers are constant source of "security" reports. My position will always be behind the strict parser, if for no reason other than to get those useless reports to stop.

bcomnes commented:
> This sounds like you are able to handle these absolutely incorrect header values.

Yes of course I can accommodate it. But I had to change in a non-breaking update. My point is that you can still tighten the validation without moving where that happens. This change was also a breaking lifecycle policy change. 

https://github.com/fastify/fastify/pull/6918 effectively keeps your change but avoids the breaking change which doesn't seem nessisary for any of the goals here unless I'm mistaken.

jsumners commented:
My goals were accomplished. The content-type parser was permitting invalid headers to pass through and/or jump to incorrect logic. This PR fixed those bugs. I am against any change that restores the previous behavior.

bcomnes commented:
The current implementation still lets some invalid headers through to addContentTypeParser. (Do you want test more cases?) Is the intention to strictly gate all of those out now? Clarifying intention here would help me understand what's acceptable here.

jsumners commented:
I have seen you mention a case where a type+subtype is appended to the parameters list. That's just a bad parameters list and this feature explicitly opts to best effort those at this time. They are not used internally except to match incoming requests to the correct parser. A patch that rejects this malformed content in an efficient manner would be acceptable.

bcomnes commented:
Just chatted with @mcollina on a call about this and have a better understanding of what the issue/constraints are. 

- The goal is to strictly drop invalid content type headers into addContentTypeParser, because the content-type narrowed schema validator can be bypassed by invalid content-type headers (kinda seems like an issue with the schema in an app implementation, but I guess its come up enough to necessitate this gate due to the validation guarantees). I may have found some invalid headers that are still getting through. Let me look into weather they matter or not for this concern. 
- Adding an option to disable the strict validation branch in into `contentTypeParser.run` in `lib/handle-request.js` is a possibility, or maybe adding a hook to swap in custom ContentParser implementation. Might be worth adding as an escape hatch due to breaking nature in a non-major release.
- Mushing around content-type headers in onRequest into something valid is fine, maybe the preferred place to do it, but `addContentTypeParser` should document about this new limitation (they don't run if they don't pass fastify's validation).
- The whatwg algo doesn't materially improve the security needs here, and the standards behavior defensibility isn't relevant for this issue.

>  A patch that rejects this malformed content in an efficient manner would be acceptable.

Will poke around if I can come up with something, thank you for the guidance.

bcomnes commented:
Small docs PR: https://github.com/fastify/fastify/pull/6925

Explains the new behavior and hazard.

bcomnes commented:
This PR: https://github.com/fastify/fastify/pull/6926

Makes strict validation prior to consulting `addContentTypeParser` optional, defaulting to the new strict behavior.

bcomnes commented:
This PR: https://github.com/fastify/fastify/pull/6927

Adds a different/alternative/complementary escape hatch allowing callers to customize/extend/replace the ContentType parser.

bcomnes commented:
I sent some time trying to work through understand what is needed here wrt strict RFC 9110 validation, but also the lax trailer parsing, and normalization recovery that remains. 


```
- \u00a0application/json
- application/json\u00a0
- application/json\u00a0; charset=utf-8
```

Looks like there might be some whitespace normalization leaks as these seems to leak through. Currently `contentType.isValid` seems to mean "Fastify recovered a usable `type/subtype`." Not: “The complete original header value conforms to the RFC 9110 `Content-Type` grammar.” Not sure what the intention here happy to expand this validation if people think its needed.

jsumners commented:
I honestly can't remember the full details around it. But my vague recollection is that parts of the codebase allowed for empty values while needing to distinguish such a value from an invalid header.

Improvements are welcome.
